File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.js
function solution(book_time) {
    const bookTime = book_time.map((b) => {
        const [h1, m1] = b[0].split(':').map(Number)
        const [h2, m2] = b[1].split(':').map(Number)
        
        return [h1*60+m1, h2*60+m2]
    }).sort((a, b) => a[0] - b[0])
    
    
    const minHeap = []
    
    function addHeap(end){
        minHeap.push(end)
        minHeap.sort((a, b) => a - b)
    }
    
    for(let [start, end] of bookTime){
        const startTime = start
        const endTime = end + 10
        
        if(minHeap.length > 0 && minHeap[0] <= startTime){
            minHeap.shift()
        }
        addHeap(endTime)
    }
    
    return minHeap.length
}

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.*;

class Solution {
    public int solution(String[][] book_time) {        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> {
            if(a[0] == b[0]) return a[1] - b[1];
            return a[0] - b[0];
        });
        
        for (String[] times : book_time) {
            int start = toMinute(times[0]);
            int end = toMinute(times[1]) + 10;

            // start, end 사용
            pq.add(new int[]{start, end});
        }
        
        
        PriorityQueue<Integer> rooms = new PriorityQueue<>();
        int max = 0;
        while(!pq.isEmpty()){
            int[] cur = pq.poll();
            
            if(rooms.isEmpty()){
                rooms.add(cur[1]);
                continue;
            }
            
            int prev = rooms.poll();
            
            if(prev <= cur[0]){
                rooms.add(cur[1]);
            } else {
                rooms.add(prev);
                rooms.add(cur[1]);
            }
            
            if(rooms.size() > max){
                max = rooms.size();
            }
            
        }
        
        return max == 0 ? 1 : max;
        
    }
    
    static int toMinute(String time) {
        String[] t = time.split(":");
        int hour = Integer.parseInt(t[0]);
        int minute = Integer.parseInt(t[1]);
        
        return hour * 60 + minute;
    }

}